scvi/model/base/_pyromixin.py

```diff
@@ -0,0 +1,593 @@
+        self,
+        args,
+        kwargs,
+        num_samples: int = 1000,
+        return_sites: Optional[list] = None,
+        return_observed: bool = False,
+        show_progress: bool = True,
+    ):
+        """
+        Get many samples from posterior distribution.
+
+        Parameters
+        ----------
+        args
+            arguments to model and guide
+        kwargs
+            arguments to model and guide
+        show_progress
+            show progress bar
+
+        Returns
+        -------
+        Dictionary with array of samples for each variable
+        dictionary {variable_name: [array with samples in 0 dimension]}
+
+        """
+
+        samples = self._get_one_posterior_sample(
+            args, kwargs, return_sites=return_sites, sample_observed=return_observed
+        )
+        samples = {k: [v] for k, v in samples.items()}
+
+        for _ in tqdm(
+            range(1, num_samples),
+            disable=not show_progress,
+            desc="Sampling global variables, sample: ",
+        ):
+            # generate new sample
+            samples_ = self._get_one_posterior_sample(
+                args, kwargs, return_sites=return_sites, sample_observed=return_observed
+            )
+
+            # add new sample
+            samples = {k: samples[k] + [samples_[k]] for k in samples.keys()}
+
+        return {k: np.array(v) for k, v in samples.items()}
+
+    def _posterior_samples_full_data(self, use_gpu: bool = True, **sample_kwargs):
+        """
+        Generate samples from posterior distribution using all data
+
+        Parameters
+        ----------
+        sample_kwargs
+            arguments to _get_posterior_samples
+
+        Returns
+        -------
+        dictionary {variable_name: [array with samples in 0 dimension]}
+
+        """
+
+        self.module.eval()
+        gpus, device = parse_use_gpu_arg(use_gpu)
+
+        args, kwargs = self.module.model._get_fn_args_full_data(self.adata)
+        args = [a.to(device) for a in args]
+        kwargs = {k: v.to(device) for k, v in kwargs.items()}
+        self.to_device(device)
+
+        samples = self._get_posterior_samples(args, kwargs, **sample_kwargs)
+
+        return samples
+
+    def _posterior_samples_minibatch(self, use_gpu: bool = True, **sample_kwargs):
+        """
+        Generate samples of the posterior distribution of each parameter, separating local (minibatch) variables
+        and global variables, which is necessary when performing minibatch inference.
+
+        Note for developers: requires model class method which lists observation/minibatch plate
+        variables (self.module.model.list_obs_plate_vars()).
+
+        Parameters
+        ----------
+        use_gpu
+            Bool, use gpu?
+
+        Returns
+        -------
+        dictionary {variable_name: [array with samples in 0 dimension]}
+
+        """
+
+        gpus, device = parse_use_gpu_arg(use_gpu)
+
+        self.module.eval()
```

Comment:
eval should be called in the training procedure. So I'd remove this

Reply:
@adamgayoso do you mean at the very end of training? e.i. after this line? https://github.com/YosefLab/scvi-tools/pull/1059/files#diff-cb30789d40b9630e34b2067718c7d61aba3238123aa1a7ae8852c80846031b6fR272